### PR TITLE
Doctests: running examples from docstings automatically

### PR DESCRIPTION
**Pull request by romovpa**
## Doctests

Issue and motivation: #229. We are going to write plenty of tests. I want to encourage the team to use doctests whenever it's practical. By doing so we feed two birds with one seed: improve documentation and test coverage.

[Doctest](https://docs.pytest.org/en/6.2.x/doctest.html) are just docstrings run as unit tests with pytest. For example:
```python
class SequenceBias(nn.Module):
    r"""...

    Example:
        >>> m = SequenceBias(16)
        >>> input = torch.randn(20, 4, 16)
        >>> output = m(input)
        >>> print(output.size())
        torch.Size([21, 4, 16])
    """
```

Pytest:
```
$ pytest --doctest-modules -p conftest opacus/layers/dp_multihead_attention.py
=========================== test session starts ===========================
platform darwin -- Python 3.9.7, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /Users/romovpa/workspace/opacus
plugins: hypothesis-6.24.2
collected 1 item                                                          

opacus/layers/dp_multihead_attention.py .                           [100%]

============================ 1 passed in 0.01s ============================
```

## Changes

- Exising doctests are fixed.
- Removed `docstring_examples_test.py`. 
    - We don't need to synchronize tests with docstrings anymore
    - Disabled PrivacyEngine tests are moved to `privacy_engine.py` as TODO 
- Circle-CI command modified

## How Has This Been Tested (if it applies)

All pytests passed (including doctests):
```
$ pytest --doctest-modules -p conftest opacus
```

## Further steps

- Write guidelines on how to use doctests efficiently
- Update CONTRIBUTING

### CODE REVIEW

**Review by ffuuugor — APPROVED** [review]
Great initiative, thanks for following through on your suggestion in the issues!

Just to clarify, what's the correct way of using code in the docstrings so it's picked up by `doctests`?

**Comment by romovpa** [comment]
> Just to clarify, what's the correct way of using code in the docstrings so it's picked up by `doctests`?

Basically, just add examples in the docstrings. Pytest will detect them automatically and run as test.

Some things you need to know:

1. **Formatting**. The code should be formatted as if you were using REPL: 
    - `>>>` means start of an expression
    - `...` means the continuation of the expression
    - The text after the expression is a correct expected output
2. **Skip**. By default all the code in docstrings will be run. If you don't want to test expression, put a comment `# doctest: +SKIP` 
3. **Output comparison**. Copy the output exactly as it is in REPL.
    - You can put `...` in the docstring's output, this will match any substring. For example `tensor([...])` will match any inline tensor. This is handy when using randomized input.
    - You can add `# doctest: +NUMBER` on the expression, this will compare numbers automatically rounding them to the precision used in the docstring.
4. **Namespace**. 
    - If you want to use a module which is not imported outside of the docstring, import it in the docstring.
    - Some frequently used modules and classes are added automatically in the test namespace, such as `torch`, `nn`, `np`, `opacus`, etc. See `conftest.py`.
    - We can add toy classes to demonstrate functions on them. For example I created `MyCustomModule` as a real `nn.Module`, it is available to use in docstring tests.

**Review by karthikprasad — APPROVED** [review]
Great initiative indeed @romovpa ! Thanks a lot for this suggestion and the PR 😃 

> Some things you need to know:

Not as part of this PR, but it would be good to add your comment to CONTRIBUTING.md file as well for developers to follow.

**Comment by romovpa** [comment]
@ffuuugor fyi, `test_noise_level` has failed (flaky)

**Comment by ffuuugor** [comment]
> @ffuuugor fyi, test_noise_level has failed (flaky)

Thanks for the heads up. That's expected, I've addressed this in one of the open PRs:

https://github.com/pytorch/opacus/pull/254/files#diff-af4fd706ae53070877bde3d1dc350c8bbbdc51db6e88b4b07748dc3a983334deR45
